# Implementando Regressão Múltipla do Zero

Nessa tarefa você vai estender sua implementação da tarefa passada para considerar múltiplas variáveis. Você pode estender a versão vetorizada implementada [neste notebook](https://canvas.instructure.com/courses/1389733/files/68372962/download?verifier=jx3MnZn2ltc955iEt69PwjkzJvBiuQyUNXwGA9gl&wrap=1) para regressão simples. 

- Rode o algoritmo [nesses dados](https://canvas.instructure.com/courses/1389733/files/68372968/download?verifier=9Qm2NF62mBNlbfeLJWfrzjGl6qvS0eIuX3kOGNAU&wrap=1), onde as linhas representam as notas de alunos de computação de alunos da UFCG em algumas disciplinas do primeiro período. A última coluna é a variável alvo representando o CRA final depois de concluir o curso. As outras colunas são algumas disciplinas do primeiro período. O pressuposto aqui é que as notas em disciplinas no primeiro período ajudam a explicar o CRA final dos alunos de computação.

- Compare o valor dos coeficientes estimados pelo seu algoritmo com o valor dos coeficientes da regressão linear do scikit learn para testar se o seu algoritmo está funcionando corretamente.

A entrega deve ser o link no seu github para o notebook Jupyter com código python e texto explicativo quando necessário. De preferência, crie um repositório na sua conta do github e envie o link do html do notebook.

## Implementação utilizando a biblioteca scikit-learn

### Com a biblioteca numpy

In [62]:
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

In [74]:
points = np.genfromtxt('sample_treino.csv', delimiter=',')

In [64]:
# Arrays x e y

In [65]:
# Parametros
X = points[1:, :-1]
y = points[1:,-1]

In [66]:
from sklearn.linear_model import LinearRegression

In [67]:
lm = LinearRegression()

In [68]:
lm.fit(X, y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [69]:
### Avaliação do modelo

In [70]:
print(lm.intercept_)

1.737711513794439


In [161]:
print('Coeficientes')
atributo = np.array(['Cálculo1','LPT','P1','IC','Cálculo2'])
for i in range(0,5):
    print(atributo[i],":",round(lm.coef_[i],6))

Coeficientes
Cálculo1 : 0.103041
LPT : 0.046437
P1 : 0.164098
IC : 0.381178
Cálculo2 : 0.020278


### Forma da regressão linear multivariada para o dataset

 $\hat{y}$ = 1.737711513794439 + 0.103041$\cdot$**Cálculo1** + 0.046437$\cdot$**LPT** + 0.164098$\cdot$**P1** + 0.381178$\cdot$**IC** + 0.020278$\cdot$**Cálculo2**

Interpretando os coeficientes:

- Mantendo todas as outras variáveis constantes, um aumento de 1 unidade em **Cálculo1** está associado a um **aumento de 0.103041** no CRA.
- Mantendo todas as outras variáveis constantes, um aumento de 1 unidade em **LPT** está associada a um **aumento de 0.046437** no CRA.
- Mantendo todas as outras variáveis constantes, um aumento de 1 unidade em **P1** está associada a um **aumento de 0.164098** no CRA.
- Mantendo todas as outras variáveis constantes, um aumento de 1 unidade em **IC** está associada a um **aumento de 0.381178** no CRA.
- Mantendo todas as outras variáveis constantes, um aumento de 1 unidade em **Cálculo2** está associado a um **aumento de 0.020278** no CRA.

### Com a biblioteca pandas

In [61]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

In [73]:
points2 = pd.read_csv('sample_treino.csv')

In [78]:
points2.columns

Index(['Cálculo1', 'LPT', 'P1', 'IC', 'Cálculo2', 'cra'], dtype='object')

In [79]:
# Parametros
X1 = points2[['Cálculo1', 'LPT', 'P1', 'IC', 'Cálculo2']]
y1 = points2['cra']

In [80]:
from sklearn.linear_model import LinearRegression

In [81]:
lm2 = LinearRegression()

In [82]:
lm2.fit(X1, y1)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [83]:
print(lm2.intercept_)

1.7377115137944363


In [162]:
coeff_points = pd.DataFrame(lm2.coef_,X1.columns,columns=['Coefficient'])
coeff_points

Coefficient
Cálculo1     0.103041
LPT          0.046437
P1           0.164098
IC           0.381178
Cálculo2     0.020278

### Forma da regressão linear multivariada para o dataset

 $\hat{y}$ = 1.737711513794439 + 0.103041$\cdot$**Cálculo1** + 0.046437$\cdot$**LPT** + 0.164098$\cdot$**P1** + 0.381178$\cdot$**IC** + 0.020278$\cdot$**Cálculo2**

Interpretando os coeficientes:

- Mantendo todas as outras variáveis constantes, um aumento de 1 unidade em **Cálculo1** está associado a um **aumento de 0.103041** no CRA.
- Mantendo todas as outras variáveis constantes, um aumento de 1 unidade em **LPT** está associada a um **aumento de 0.046437** no CRA.
- Mantendo todas as outras variáveis constantes, um aumento de 1 unidade em **P1** está associada a um **aumento de 0.164098** no CRA.
- Mantendo todas as outras variáveis constantes, um aumento de 1 unidade em **IC** está associada a um **aumento de 0.381178** no CRA.
- Mantendo todas as outras variáveis constantes, um aumento de 1 unidade em **Cálculo2** está associado a um **aumento de 0.020278** no CRA.